pkg/cmd/predict.go

```diff
@@ -0,0 +1,289 @@
+package cmd
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/kubecost/kubectl-cost/pkg/query"
+
+	"github.com/opencost/opencost/pkg/log"
+
+	"github.com/spf13/cobra"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/runtime"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/rest"
+
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// PredictOptions contains options specific to prediction queries.
+type PredictOptions struct {
+	window string
+
+	clusterID string
+
+	// The file containing the workload definition to be predicted.
+	filepath string
+
+	showCostPerResourceHr bool
+
+	query.QueryBackendOptions
+}
+
+func newCmdPredict(
+	streams genericclioptions.IOStreams,
+) *cobra.Command {
+	kubeO := NewKubeOptions(streams)
+	predictO := &PredictOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "predict",
+		Short: "Estimate the monthly cost of a workload based on tracked cluster resource costs.",
+		RunE: func(c *cobra.Command, args []string) error {
+			if err := kubeO.Complete(c, args); err != nil {
+				return fmt.Errorf("complete k8s options: %s", err)
+			}
+			if err := kubeO.Validate(); err != nil {
+				return fmt.Errorf("validate k8s options: %s", err)
+			}
+
+			if err := predictO.Complete(kubeO.restConfig); err != nil {
+				return fmt.Errorf("complete: %s", err)
+			}
+			if err := predictO.Validate(); err != nil {
+				return fmt.Errorf("validate: %s", err)
+			}
+
+			return runCostPredict(kubeO, predictO)
+		},
+	}
+	cmd.Flags().StringVarP(&predictO.filepath, "filepath", "f", "", "The file containing the workload definition whose cost should be predicted. E.g. a file might be 'test-deployment.yaml' containing an apps/v1 Deployment definition. '-' can also be passed, in which case workload definitions will be read from stdin.")
+	cmd.Flags().StringVarP(&predictO.clusterID, "cluster-id", "c", "", "The cluster ID (in Kubecost) of the presumed cluster which the workload will be deployed to. This is used to determine resource costs. Defaults to all clusters.")
+	cmd.Flags().BoolVar(&predictO.showCostPerResourceHr, "show-cost-per-resource-hr", false, "Show the calculated cost per resource-hr (e.g. $/byte-hour) used for the cost prediction.")
```

Comment:
I'm just using this code review as an opportunity to learn kubectl predict feature i believe Niko is on this for better  review than me, is there a reason why monthly CPU/memory cost is returned while hourly cost is based on a flag?

Reply:
`kubectl cost` has a pattern of only showing basic, but high value information by default. You can then add flags to show a more complex picture, as desired. See e.g. `kubectl cost namespace -A` https://github.com/kubecost/kubectl-cost/blob/a5864b0d578ef01a3c3ea033e9d55c123cb76f79/pkg/cmd/common.go#L50-L59

So, here, we by default just show the predicted monthly cost for the workload, broken up by resource type. If you want to introspect _why_ what is the cost, you use `--show-cost-per-resource-hour` to have an explanation of what cost basis Kubecost used for the prediction, e.g. what cost per cpu-core hour or cost per byte hour or cost per gpu-hour was used to arrive at the given cost.

My gut says that users aren't interested in this context by default, but I also think it is helpful to have available if desired.